File: sample/src/main/java/tech/thdev/androidrecyclerview/view/animation/AnimatorEventSubScribe.java
package tech.thdev.androidrecyclerview.view.animation;

import android.animation.Animator;
import android.util.Log;

import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.disposables.Disposable;

/**
 * Created by Tae-hwan on 20/10/2016.
 */
public class AnimatorEventSubScribe implements ObservableOnSubscribe<ItemAnimatorEvent> {

    private Animator animator;

    public AnimatorEventSubScribe(Animator animator) {
        this.animator = animator;
    }

    @Override
    public void subscribe(final ObservableEmitter<ItemAnimatorEvent> e) throws Exception {
        final Animator.AnimatorListener listener = new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
                if (!e.isDisposed()) {
                    e.onNext(ItemAnimatorEvent.eventCreate(animation, ItemAnimatorEvent.ANIM_START));
                }
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                if (!e.isDisposed()) {
                    e.onNext(ItemAnimatorEvent.eventCreate(animation, ItemAnimatorEvent.ANIM_END));
                }
            }

            @Override
            public void onAnimationCancel(Animator animation) {

            }

            @Override
            public void onAnimationRepeat(Animator animation) {

            }
        };

        animator.addListener(listener);
        animator.start();

        e.setDisposable(new Disposable() {
            @Override
            public void dispose() {
                animator.cancel();
                animator.removeAllListeners();
            }

            @Override
            public boolean isDisposed() {
                return true;
            }
        });
    }
}

File: sample/src/main/java/tech/thdev/androidrecyclerview/ui/animation/AnimatorEventSubScribe.kt
package tech.thdev.androidrecyclerview.ui.animation

import android.animation.Animator
import io.reactivex.rxjava3.core.FlowableEmitter
import io.reactivex.rxjava3.core.FlowableOnSubscribe
import io.reactivex.rxjava3.disposables.Disposable

/**
 * Created by Tae-hwan on 20/10/2016.
 */
class AnimatorEventSubScribe(private val animator: Animator) :
    FlowableOnSubscribe<ItemAnimatorEvent> {

    @Throws(Exception::class)
    override fun subscribe(emitter: FlowableEmitter<ItemAnimatorEvent>?) {
        val listener = object : Animator.AnimatorListener {
            override fun onAnimationRepeat(animation: Animator?) = Unit

            override fun onAnimationEnd(animation: Animator?) {
                if (emitter?.isCancelled != true && animation != null) {
                    emitter?.onNext(
                        ItemAnimatorEvent.eventCreate(
                            animation,
                            ItemAnimatorEvent.ANIM_END
                        )
                    )
                }
            }

            override fun onAnimationCancel(animation: Animator?) = Unit

            override fun onAnimationStart(animation: Animator?) {
                if (emitter?.isCancelled != true && animation != null) {
                    emitter?.onNext(
                        ItemAnimatorEvent.eventCreate(
                            animation,
                            ItemAnimatorEvent.ANIM_START
                        )
                    )
                }
            }
        }

        animator.addListener(listener)
        animator.start()
        emitter?.setDisposable(object : Disposable {
            override fun isDisposed(): Boolean = true

            override fun dispose() {
                animator.cancel()
                animator.removeAllListeners()
            }
        })
    }
}